package/MDAnalysis/core/topologyobjects.py

```diff
@@ -107,9 +107,14 @@ def type(self):
     def is_guessed(self):
         return bool(self._guessed)
 
+    # TODO: which one to use?
```

Comment:
Everything in core/test_topologyobjects passes with both functions.

These hash functions behave differently. This is because the `_cmp_key` can return the same value for two different TopologyObjects. 

```python
    def _cmp_key(self):
        """Unique key for the object to be used to generate the object hash"""
        # This key must be equal for two object considered as equal by __eq__
        return self.__class__, tuple(sorted(self.indices))
```

because we to a sorting of the indices a object with indices (2, 1, 3) will have the same key as an object with (3, 1, 2) or any other permutation of these 3 values. It also only compares the indices independent from the universe this object is attached to.

The first hash function is a little bit better here since it compares against the universe and preserves the order of indices.